Variables support

gheon opened the pull request:
This feature enables end-users to use ProActive workflow/task variables inside the kubernetes manifest file.

Commit:
Add support for PA variables inside k8s manifest.

Commit:
Remove old unused imports.

Mykhailenko commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -140,6 +140,16 @@ else if (k8sDeleteOnly) {
         try {
+
+            for (Map.Entry<String, String> variable : variables.entrySet()) {
+                if (k8s_manifest.contains("${variables_" + variable.getKey() + "}")) {
```
I believe Java Code Convention does not recommend to use such_type_of_names.
But as you want I do not insist.

Mykhailenko commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -140,6 +140,16 @@ else if (k8sDeleteOnly) {
+
             k8sManifestFile = new GenericFileWriter().forceFileToDisk(k8s_manifest, K8S_MANIFEST_FILE_NAME);
         } catch (IOException e) {
             log.warn("Failed to write content to kubernetes manifest file: ", e);
```
Are you sure that warn level is sufficient here?

gheon replied:
Right, I set it to error as of now.

Mykhailenko commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -140,6 +140,16 @@ else if (k8sDeleteOnly) {
+
+            for (Map.Entry<String, String> variable : variables.entrySet()) {
+                if (k8s_manifest.contains("${variables_" + variable.getKey() + "}")) {
+                    k8s_manifest = org.apache.commons.lang3.StringUtils.replace(k8s_manifest,
```
I prefer static imports instead of full qualifiers..

fviale commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -347,6 +358,10 @@ private void populateBindingAndGI() throws ScriptException {
+
+        // Retrieving passed variables if any
+        variables = (Map<String, String>) context.getBindings(ScriptContext.ENGINE_SCOPE)
+                                                 .get(SchedulerConstants.VARIABLES_BINDING_NAME);
```
instead of only getting the variables and generic information bindings, it is more generic to simply iterate existing bindings and depending on the type, build a replacement map. 
Example:
1) if binding is a map:
for each key inside the map, ${bindingName_key} => value.toString()
2) if binding is a list, array, collection(?)
for each element, ${bindingName_index} => value.toString()
3) otherwise
${bindingName} => value.toString()

fviale replied:
ref https://github.com/ow2-proactive/jsr223-nativeshell/blob/master/src/main/java/jsr223/nativeshell/NativeShellRunner.java

Commit:
Better handling of bindings (variables, GI, env).

Commit:
Fix references to GI for engine custom behavior.

Commit:
Upgrade version number.

Commit:
Remove unused log.

fviale commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -133,13 +129,27 @@ else if (k8sDeleteOnly) {
+        Map<String, String> environment = bindings.getEnvironment();
         try {
+            for (Map.Entry<String, String> variable : environment.entrySet()) {
+                if (k8s_manifest.contains("${" + variable.getKey() + "}")) {
```
you can define a local variable to contain the pattern

fviale replied:
alternatively you can put in the replacement map directly the correct pattern

gheon replied:
Right, done.

fviale commented on src/main/java/jsr223/kubernetes/utils/BindingUtils.java:
```diff
@@ -0,0 +1,91 @@
+    @Getter
+    Map<String, String> environment;
+
+    private void addMapBindingAsEnvironmentVariable(String bindingKey, Map<?, ?> bindingValue,
```
please rename these methods, for example to addMapBindingToReplacementMap(String bindingKey, Map<?, ?> mapBindingValue, Map<String, String> replacementMap)

etc...

gheon replied:
The thing is, not only replacements var will be extracted from the bindings - there is also others GI used to customize the behavior.  That's why I kept originally the 'environment' name.

gheon replied:
Eventually renamed to addMapBindingToEngineMetadata, etc.

fviale replied:
ok why not

fviale commented on src/main/java/jsr223/kubernetes/utils/BindingUtils.java:
```diff
@@ -0,0 +1,91 @@
+
+    // GI, bindings and variables
+    @Getter
+    Map<String, String> environment;
```
rename this to replacementMap

gheon replied:
renamed to k8sEngineMetadata

fviale commented on src/main/java/jsr223/kubernetes/KubernetesScriptEngine.java:
```diff
@@ -133,13 +129,27 @@ else if (k8sDeleteOnly) {
+            for (Map.Entry<String, String> variable : environment.entrySet()) {
+                if (k8s_manifest.contains("${" + variable.getKey() + "}")) {
+                    k8s_manifest = org.apache.commons.lang3.StringUtils.replace(k8s_manifest,
+                                                                                "${" + variable.getKey() + "}",
```
you can eventually copy/paste, adapt and use the method VariableSubstitutor.replaceRecursively().

This allows recursive replacements such as ${variables_PARAM_${variables_PA_TASK_ITERATION}}

Commit:
Add recursive wf/task variables substitution.

Commit:
Improve bindings handling.

- Use final string for string constant
- Rename from environment -> engine metadata.

Commit:
Make use of recursive variables substitutor.

gheon commented:
@fviale @Mykhailenko thanks for your feedbacks, I have improved the code accordingly. If you are okay, we can merge.

fviale commented on src/main/java/jsr223/kubernetes/utils/VariablesSubstitutor.java:
```diff
@@ -0,0 +1,62 @@
+                                                      K8S_MANIFEST_CUSTOM_VAR_SYNTAX_SUFF, variable.getValue());
+                    anyReplacement = anyReplacement || !newOutput.equals(output);
+                    output = newOutput;
+                    System.out.println("I found.");
```
please remove this System.out :-)

fviale reviewed (approved):
One important comment, otherwise looks good

Commit:
Remove System.out.println call.
